docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,228 @@
+
+###### Lease data structure for a shoot
+
+* Owning seed cluster
+* Lease Period
+  * Lease start timestamp
+  * Lease end timestamp or TTL
+
+DNS seems well suited to address this problem of lease service. We can use DNS records with short TTL to
+publish/broadcast the owning seed for any particular shoot. DNS is highly available and the short TTL can act like
+the lease. Garden shoot-controller or some additional controller can populate or renew the DNS records based on the 
+shoot resource records in the garden cluster.
+
+##### Shoot Control-plane Co-ordinator
+
+The Shoot Control-plane Co-ordinator is a pod (actually, a deployment) deployed as a part of each of the shoot control-plane in the seed clusters.
+
+It has the responsibility to monitor whether the current seed cluster where it is deployed has the lease for the given shoot using any of the mechanisms chosen [above](#advertising-current-seed-cluster-lease-for-a-shoot-cluster) and take appropriate action to activate/passivate the relevant shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc.
+
+It has two states.
+
+1. Active
+
+In the active state, the co-ordinator is able to reach the lease service and knows that the current seed is the owner for the given shoot according to the lease service.
+
+In such a scenario, the co-ordinator makes sure the shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc. are scaled up and active.
+
+2. Passive
+
+In the passive state, the co-ordinator is either not able reach the lease service or knows that the current seed is not the owner of the given shoot according to the lease service.
+
+In such a scenario, the co-ordinator passivates the shoot control-plane components such as machine-controller-manager, etcd statefulset, VPN etc. by scaling them down to zero.
+
+It should not scale itself down, of course.
+
+##### State Diagram
+
+![State Diagram](state-diagram.png)
+
+This component should also expose a `/healthz` endpoint which returns healthy only if it can get the current lease status and if the current seed cluster is the owner of the lease for the given shoot. This `/healthz` endoint can be used for livenessProbe in the shoot machine-controller-manager, shoot etcd and also perhaps in the VPN components to avoid co-ordination problems in case the shoot control-plane co-ordinator itself crashes.
+
+###### Melt-down scenario
+
+In the [above](#shoot-control-plane-co-ordinator) approach, there is a possibility of a melt-down-like scenario where the co-ordinator is not able to access the lease status for some long period even though it actually owns the lease during that period, but the gardener is able to access the seed cluster. In such a scenario, co-ordinator will scale down the relevant component to zero while a gardener reconciliation will scale them back up to the original values. This cycle can potentially keep happening while the scenario persists.
```

Comment:
As said, I would be interested to explore the possibility of acting only if the lease is definitely not hold by this seed cluster and avoid actions when the lease service is unavailable or the coordinator is failing. This way, we run less risk of scaling a healthy control plane down and we also wouldn't have to deal with the complex corner cases.

The question is really, whether something would totally break in this scenario and whether we could mitigate that specifically.

Reply:
We can mitigate this individually for etcd and mcm separately.
1. etcd-backup-restore can use the bucket itself to update ownership or lease without using even DNS or any other kind of lease service.
2. mcm will have a harder time but perhaps it can use tags on the vms to update ownership or lease as well?

The advantage here would be that we won't have false positives due to network/DNS downtime.

The disadvantage is that each sensitive control-plane component will have to solve this problem individually. Any new component also will have to solve this problem itself.

Reply:
BTW we can keep the DNS TTL relatively long as well. As long as it is faster than manual operator response :-)

This can be another way to mitigate false positives. But at the risk of proportionately delaying the new seed movement from taking effect.

Reply:
Well, in regards to etcd-backup-restore, the solution could be to simply copy the bucket and configure the moved control plane to use the cloned bucket and delete the old bucket. Gardener is in control and can do that when moving the control plane (it also changes the DNS record).

The other outbound control is with the machines, yes. Here we simply don't need to look into corner cases, because no harm can be done and eventually the DNS record will switch.

The last outbound control runs over the API server through VPN into the shoot, but this can be solved via the new API server and the VPN shoot pod, e.g. by redeploying it with changed credentials.

As for the TTL, it must be short, not only because of above, but also because of the workers/`kubelet`s that must see the new control plane API server fast.

The next advantage, if we do not build a "control plane terminator", is that this way the new control plane can start right away. The moment the DNS entry sits (regular provisioning), the rest of the control plane gets deployed and starts immediately (present flow logic). When the Gardener moves the control plane, it clones the backup and changes the DNS entry. When it sits, it waits TTL seconds before deploying the rest (including machine controller manager) and everything starts as well.

Anyway, it's an idea, maybe worth considering? What I don't like about my proposal is it specificity. On the other hand, it is less effort, has no problematic corner cases (hopefully), needs less code, works fast.

Reply:
Even easier, backup & restore credentials could be changed.

Reply:
Sure. But if we are thinking along these lines, I think just leaving two MCM instances for some time (and if we do not have some lease mechanism such time can be indefinitely long), is equally problematic especially with auto-scaler switched on.

If we are thinking of specific solutions for the individual components, I think we should solve it within MCM without relying on the possibility that not too many things would change before the gardener is able to reconcile/delete the older control-plane.

But before we switch to the specific solution approach, I think it might be worth checking how much of a problem DNS being unavailable temporarily really is.

Reply:
Again, can't we change the credentials (now for IaaS/MCM) and isn't that the general cure to implement all this (control plane migration) safely?

This way, when Gardener decides to migrate a control plane it:
1. Switches the DNS entry redirecting **inbound** communication away from the decommissioned control plane and to the new control plane
2. But it also rotates the credentials, disabling **outbound** communication from the decommissioned control plane

Not only do we have to implement that anyway eventually (certificate/secret revocation and rotation), we would make positive use of it right away for our scenario and completely dissolve the "problem" safely. No need to make the decommissioned control plane shut down involuntarily (and risk losing also healthy control planes due to bugs), no complex corner cases, we use only basic mechanisms to achieve our goals (backup & restore/recovery, revoke permissions, DNS redirection).

Reply:
Changing MCM credentials was @mvladev's suggestion in the review meeting. But this is not possible, as @rfranzke mentioned in the same meeting, because the MCM credentials are the credentials supplied by the shoot owner which cannot be changed by Gardener.

ETCD credentials can be changed because they are created by Gardener internally.

Reply:
Summarising the discussion during the weekly sync...

We can think of changing the gardener to create new credentials for shoot mcm. This can make the revoking access approach workable for both etcd and mcm.

#### Pros
1. Gardener then requires only identity and access management access in the shoot credentials.
1. Gardener can rotate credentials freely. This can enhance security of the seed and shoot clusters.
#### Cons
1. This approach may or may not generalise to other shoot control-plane components.

We will need to use DNS still to make sure the existing nodes can reach the new shoot apiserver.

#### Action Items
1. Need to evaluate the size of the change required in gardener.
1. Need to evaluate this approach for some of the upcoming additional shoot control-plane components.